enterprise/server/sociartifactstore/sociartifactstore.go

```diff
@@ -4,52 +4,79 @@
 package sociartifactstore
 
 import (
+	"bytes"
 	"context"
 	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
 	"net/http"
 	"os"
 	"strconv"
 	"strings"
 	"time"
 
+	"github.com/awslabs/soci-snapshotter/compression"
 	"github.com/awslabs/soci-snapshotter/soci"
-	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/commandutil"
+	"github.com/awslabs/soci-snapshotter/ztoc"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/util/dsingleflight"
 	"github.com/buildbuddy-io/buildbuddy/server/environment"
 	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
 	"github.com/buildbuddy-io/buildbuddy/server/util/log"
 	"github.com/buildbuddy-io/buildbuddy/server/util/prefix"
 	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/containerd/containerd/images"
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/google/go-containerregistry/pkg/v1/match"
 	"github.com/google/go-containerregistry/pkg/v1/partial"
 	"github.com/google/go-containerregistry/pkg/v1/remote"
 	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
 	"github.com/google/go-containerregistry/pkg/v1/types"
 	"golang.org/x/sync/errgroup"
+	"google.golang.org/protobuf/proto"
 
 	rgpb "github.com/buildbuddy-io/buildbuddy/proto/registry"
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
 	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
 	socipb "github.com/buildbuddy-io/buildbuddy/proto/soci"
 	ctrname "github.com/google/go-containerregistry/pkg/name"
 	v1 "github.com/google/go-containerregistry/pkg/v1"
+	godigest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
-var (
+const (
 	// Prefix for soci artifacts store in blobstore
 	blobKeyPrefix = "soci-index-"
 
-	// The directory where soci artifact blobs are stored
-	blobDirectory = "/var/lib/soci-snapshotter-grpc/content/blobs/sha256/"
+	// The span sized used to genereate the ZToCs. The ZToC generating code
+	// will create access points into the ZToC approximately every this many
+	// bytes.
+	ztocSpanSize = 1 << 22 // about 4MB
 
-	// The name of the soci artifact database file
-	sociDbPath = "/var/lib/soci-snapshotter-grpc/artifacts.db"
+	// The minimum layer size for generating ZToCs. Layers smaller than this won't be
+	ztocMinLayerSize = 10 << 20 // about 10MB
+
+	// The SOCI Index build tool. !!! WARNING !!! This is embedded in both the
+	// ZToCs and SOCI index, so changing it will invalidate all previously
+	// generated and stored soci artifacts, forcing a re-pull and re-index.
+	buildToolIdentifier = "AWS SOCI CLI v0.1"
```

Comment:
Should we change this to our own identifier?

Reply:
Yup, switched to "BuildBuddy SOCI Artifact Store v0.1"